File: android/app/src/main/java/com/vaultsync/data/sync/RetryPolicy.kt
package com.vaultsync.data.sync

import kotlin.math.min

class RetryPolicy(
    val baseDelayMs: Long = 1000L,
    val maxDelayMs: Long = 8000L,
    val maxRetries: Int = 3
) {
    fun shouldRetry(attempt: Int): Boolean {
        return attempt < maxRetries
    }

    fun calculateDelayMs(attempt: Int): Long {
        if (attempt <= 0) return baseDelayMs
        val factor = 1L shl attempt
        return min(baseDelayMs * factor, maxDelayMs)
    }
}

File: ios/Sources/VaultSync/Data/Sync/RetryPolicy.swift
import Foundation

public struct RetryPolicy: Sendable {
    public let baseDelaySeconds: Double
    public let maxDelaySeconds: Double
    public let maxRetries: Int

    public init(baseDelaySeconds: Double = 1.0, maxDelaySeconds: Double = 8.0, maxRetries: Int = 3) {
        self.baseDelaySeconds = baseDelaySeconds
        self.maxDelaySeconds = maxDelaySeconds
        self.maxRetries = maxRetries
    }

    public func shouldRetry(attempt: Int) -> Bool {
        attempt < maxRetries
    }

    public func calculateDelay(attempt: Int) -> Double {
        if attempt <= 0 { return baseDelaySeconds }
        let factor = Double(1 << attempt)
        return min(baseDelaySeconds * factor, maxDelaySeconds)
    }
}
